GVRf/Framework/src/org/gearvrf/animation/keyframe/GVRAnimationChannel.java

```diff
@@ -0,0 +1,423 @@
+ * original transformation matrix at a specific time.<p>
+ * 
+ * This means all keys are absolute and not relative to the bone default pose.
+ * The order in which the transformations are applied is - as usual - 
+ * scaling, rotation, translation.<p>
+ */
+public final class GVRAnimationChannel implements PrettyPrint {
+    private static final String TAG = GVRAnimationChannel.class.getSimpleName();
+    protected static interface ValueInterpolator<T> {
+        T interpolate(Object begin, Object end, float factor);
+    }
+
+    protected static ValueInterpolator<Vector3f> sInterpolatorVector3f = new ValueInterpolator<Vector3f>() {
+        public Vector3f interpolate(Object begin_, Object end_, float factor) {
+            Vector3f begin = (Vector3f) begin_;
+            Vector3f end = (Vector3f) end_;
+
+            return new Vector3f().set(end).sub(begin).mul(factor).add(begin);
+        }
+    };
+
+    protected static ValueInterpolator<Quaternionf> sInterpolatorQuaternion = new ValueInterpolator<Quaternionf>() {
+        public Quaternionf interpolate(Object begin_, Object end_, float factor) {
+            Quaternionf begin = (Quaternionf) begin_;
+            Quaternionf end = (Quaternionf) end_;
+
+            return new Quaternionf().set(begin).slerp(end, factor);
+        }
+    };
+
+    protected class KeyFrameInterplator<T> {
+        GVRKeyFrame<T>[] keys;
+        ValueInterpolator<T> interpolator;
+
+        private int lastKeyIndex;
+
+        KeyFrameInterplator(GVRKeyFrame<T>[] keys, ValueInterpolator<T> interpolator) {
+            this.keys = keys;
+            this.interpolator = interpolator;
+            lastKeyIndex = -1;
+        }
+
+        protected T interpolate(float time) {
+            int index = getKeyIndex(time);
+            int nextIndex = index + 1;
+
+            if (index != -1 && keys[index].getTime() <= time && time < keys[nextIndex].getTime()) {
+                // interpolate
+                float deltaTime = (float)(keys[nextIndex].getTime() - keys[index].getTime());
+                float factor = (float)((time - keys[index].getTime()) / deltaTime);
+
+                Object start = keys[index].getValue();
+                Object end = keys[nextIndex].getValue();
+
+                return interpolator.interpolate(start, end, factor);
+            } else {
+                // time is out of range of animation time frame
+                float firstFrameTime = keys[0].getTime();
+                float lastFrameTime = keys[keys.length - 1].getTime();
+                T firstFrameValue = keys[0].getValue();
+                T lastFrameValue = keys[keys.length - 1].getValue();
+
+                if (time <= firstFrameTime) {
+                    return firstFrameValue;
+                } else if (time >= lastFrameTime) {
+                    return lastFrameValue;
+                } else {
+                    // Shouldn't happen
```

Comment:
For whatever it is worth, it crashes for me because of the RuntimeException that used to be there. No other crashes whatsoever. Also the times are consistently "sane". At the time of the crash, the "time" is always something very close to "lastFrameTime" (232.0); I get "time" of 231.71883, 231.95538, 231.41621 when the crash occurs.

Anyway, it is not going to crash anymore so if you are happy with this resolution, it is ok with me.